### DISCUSSION TITLE
Why add_template() has a references only parameters?

### DISCUSSION THREAD

**Discussion by donatengit** [mitsuhiko/minijinja#127] [Ideas]
Hello,

Thanks for all your work.

_Environment::add_template(&mut self, name: &'source str, source: &'source str)_ has only referenced parameters and that makes dynamic loading of templates based on some config ('template name'-> 'template filename') a real pain since Rust requires these references to be always valid, thus need to be stored somewhere.

Does it make sense to add implementation supporting move of Strings inside, e.g. _add_template(name: String, source: String)_ as those won't be required anywhere else outside the minijinja?

**Comment by mitsuhiko** [comment]
Have you seen the [Source](https://docs.rs/minijinja/latest/minijinja/struct.Source.html) type? It lets you load templates dynamically without running into these lifetime issues.

The short about why these lifetimes exist is that it allows the compiler and vm internally to borrow out of the loaded source efficiently. To help with this complexity the `Source` type is added that allows feeding `String` values into it.

**Reply by donatengit** [reply]
No, I haven't. Let me check. Thanks and apologies for disturbance

**Reply by mitsuhiko** [reply]
Would be good to also know how to make this more discoverable. `add_template` tries to steer users to `Source` when the bounds don't work.

**Reply by donatengit** [reply]
@mitsuhiko 

> Would be good to also know how to make this more discoverable. add_template tries to steer users to Source when the bounds don't work.

Indeed it does. But I used to rely on the very first library's readme to start. Usage of `Source` was not obvious but I should have checked examples folder.

Anyway I was able to achieve what I wanted with the library, thanks!
